### PR TITLE
Homework 21-22

### PR DESCRIPTION
**Pull request by ezlotnikova**

### CODE REVIEW

File: repository-module/src/main/java/com/gmail/ezlotnikova/repository/ShopRepository.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.gmail.ezlotnikova.repository;
+
+import java.util.List;
+
+import com.gmail.ezlotnikova.repository.model.Shop;
+
+public interface ShopRepository extends GenericRepository<Long, Shop> {
+
+    List<Shop> findShopsWithItem(Long id);
```
Comment (artemijeperevoznikov) [inline repository-module/src/main/java/com/gmail/ezlotnikova/repository/ShopRepository.java]: No need to use custom method

File: repository-module/src/main/java/com/gmail/ezlotnikova/repository/impl/ShopRepositoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    public ShopRepositoryImpl(ItemRepository itemRepository) {
+        this.itemRepository = itemRepository;
+    }
+
+    @Override
+    public List<Shop> findShopsWithItem(Long id) {
+        Item item = itemRepository.findById(id);
+        return findAll().stream()
+                .filter(shop -> shop.getItems().contains(item))
+                .collect(Collectors.toList());
```
Comment (artemijeperevoznikov) [inline repository-module/src/main/java/com/gmail/ezlotnikova/repository/impl/ShopRepositoryImpl.java]: Bad practice

File: service-module/src/main/java/com/gmail/ezlotnikova/service/converter/impl/ShopConverterImpl.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    public Shop convertDTOtoDatabaseObject(ShopDTO shopDTO) {
+        Shop shop = new Shop();
+        if (shopDTO.getId() != null) {
+            shop.setId(shopDTO.getId());
+        }
+        String name = shopDTO.getName();
+        shop.setName(name);
+        String location = shopDTO.getLocation();
+        shop.setLocation(location);
+        if (shopDTO.getItems() != null) {
```
Comment (artemijeperevoznikov) [inline service-module/src/main/java/com/gmail/ezlotnikova/service/converter/impl/ShopConverterImpl.java]: Better to use find by id

File: service-module/src/main/java/com/gmail/ezlotnikova/service/model/ShopDTO.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import javax.validation.constraints.Size;
+
+public class ShopDTO {
+
+    private Long id;
+    @Size(min = 5, max = 40, message = "Must be between 5 and 40 characters long")
+    private String name;
+    @Size(min = 5, max = 100, message = "Must be between 5 and 100 characters long")
+    private String location;
+    private Set<ItemDTO> items;
```
Comment (artemijeperevoznikov) [inline service-module/src/main/java/com/gmail/ezlotnikova/service/model/ShopDTO.java]: Better to use initial default vaue

File: spring-boot-module/src/main/java/com/gmail/ezlotnikova/springboot/controller/AppController.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        model.addAttribute("shop", new ShopDTO());
+        return "add-shop";
+    }
+
+    @PostMapping("/shops/add")
+    public String addShop(
+            @Valid @ModelAttribute(name = "shop") ShopDTO shop,
+            BindingResult errors) {
+        if (errors.hasErrors()) {
+            return "add-shop";
```
Comment (artemijeperevoznikov) [inline spring-boot-module/src/main/java/com/gmail/ezlotnikova/springboot/controller/AppController.java]: better to use underscore

File: spring-boot-module/src/main/java/com/gmail/ezlotnikova/springboot/controller/AppController.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    public String showAddItemForm(Model model) {
+        model.addAttribute("item", new ItemDTO());
+        List<ShopDTO> shops = shopService.findAll();
+        model.addAttribute("shops", shops);
+        return "add-item";
+    }
+
+    @PostMapping("/items/add")
+    public String addItem(
+            @RequestParam List<String> selectedShopIds,
```
Comment (artemijeperevoznikov) [inline spring-boot-module/src/main/java/com/gmail/ezlotnikova/springboot/controller/AppController.java]: Better to use the field at model

**Review by artemijeperevoznikov — CHANGES_REQUESTED** [review]

File: repository-module/src/main/java/com/gmail/ezlotnikova/repository/ShopRepository.java (hunk 1 above)
Comment (ezlotnikova) [reply repository-module/src/main/java/com/gmail/ezlotnikova/repository/ShopRepository.java]: Homework was re-writed wtih new unidirectional many-to-many relationship (from item to shop). All unnesesary methods and implementations deleted.

**Pull request merged**